File: lib/csharp/microsoft.bot.builder.skills/Microsoft.Bot.Builder.Skills/SkillMiddleware.cs
using System.Threading;
using System.Threading.Tasks;
using Microsoft.Bot.Builder.Dialogs;
using Microsoft.Bot.Builder.Solutions.Skills.Models;
using Microsoft.Bot.Schema;

namespace Microsoft.Bot.Builder.Solutions.Skills
{
    /// <summary>
    /// The Skill middleware is responsible for processing Skill mode specifics, for example the skillBegin event used to signal the start of a skill conversation.
    /// </summary>
    public class SkillMiddleware : IMiddleware
    {
        private readonly ConversationState _conversationState;
        private readonly IStatePropertyAccessor<DialogState> _dialogState;

        public SkillMiddleware(ConversationState conversationState, IStatePropertyAccessor<DialogState> dialogState)
        {
            _conversationState = conversationState;
            _dialogState = dialogState;
        }

        public async Task OnTurnAsync(ITurnContext turnContext, NextDelegate next, CancellationToken cancellationToken = default)
        {
            var activity = turnContext.Activity;
            if (activity != null && activity.Type == ActivityTypes.Event)
            {
                if (activity.Name == SkillEvents.CancelAllSkillDialogsEventName)
                {
                    // when skill receives a CancelAllSkillDialogsEvent, clear the dialog stack and short-circuit
                    var currentConversation = await _dialogState.GetAsync(turnContext, () => new DialogState(), cancellationToken).ConfigureAwait(false);
					if (currentConversation.DialogStack != null)
					{
						currentConversation.DialogStack.Clear();
						await _dialogState.SetAsync(turnContext, currentConversation, cancellationToken).ConfigureAwait(false);
						await _conversationState.SaveChangesAsync(turnContext, true, cancellationToken).ConfigureAwait(false);
					}

                    return;
                }
            }

            await next(cancellationToken).ConfigureAwait(false);
        }
    }
}

File: libraries/typescript/botbuilder-skills/src/skillMiddleware.ts
/**
 * Copyright(c) Microsoft Corporation.All rights reserved.
 * Licensed under the MIT License.
 */

import { ConversationState, Middleware, StatePropertyAccessor, TurnContext } from 'botbuilder';
import { DialogState } from 'botbuilder-dialogs';
import { Activity, ActivityTypes } from 'botframework-schema';
import { SkillEvents } from './models';

/**
 * The Skill middleware is responsible for processing Skill mode specifics,
 * for example the skillBegin event used to signal the start of a skill conversation.
 */
export class SkillMiddleware implements Middleware {
    private readonly conversationState: ConversationState;
    private readonly dialogStateAccessor: StatePropertyAccessor<DialogState>;

    public constructor(
        conversationState: ConversationState,
        dialogStateAccessor: StatePropertyAccessor<DialogState>
    ) {
        this.conversationState = conversationState;
        this.dialogStateAccessor = dialogStateAccessor;
    }

    public async onTurn(turnContext: TurnContext, next: () => Promise<void>): Promise<void> {
        const activity: Activity = turnContext.activity;

        if (activity !== undefined && activity.type === ActivityTypes.Event) {
            if (activity.name === SkillEvents.cancelAllSkillDialogsEventName) {

                // when skill receives a CancelAllSkillDialogsEvent, clear the dialog stack and short-circuit
                const currentConversation: DialogState|undefined = await this.dialogStateAccessor.get(turnContext, { dialogStack: [] });
                if (currentConversation !== undefined) {
                    currentConversation.dialogStack = [];
                    await this.dialogStateAccessor.set(turnContext, currentConversation);
                    await this.conversationState.saveChanges(turnContext, true);
                }

                return;
            }
        }

        await next();
    }
}
